barretenberg/cpp/src/barretenberg/ultra_honk/decider_proving_key.cpp

```diff
@@ -80,6 +80,110 @@ void DeciderProvingKey_<Flavor>::construct_databus_polynomials(Circuit& circuit)
     }
 }
 
+/**
+ * @brief Check that the number of gates in each block does not exceed its fixed capacity. Move any overflow to the
+ * overflow block.
+ * @details Using a structured trace (fixed capcity for each gate type) optimizes the efficiency of folding. However,
+ * to accommodate circuits which cannot fit into a prescribed trace, gates which overflow their corresponding block are
+ * placed into an overflow block which can contain arbitrary gate types.
+ * @note One sublety is that gates at row i may in general utilize the values at row i+1 via shifts. If the last row in
+ * a full-capacity block is such a gate, then moving the overflow out of sequence will cause that gate not to be
+ * satisfied. To avoid this, when a block overflows, the final gate in the block is duplicated, once as a dummy gate in
+ * the main block (so that the prior gate can read into it but it does not itself try to read into the next row) and
+ * again as a normal gate in the overflow block. Therefore, the total number of gates in the circuit increases by one
+ * for each block that overflows.
+ *
+ * @tparam Flavor
+ * @param circuit
+ */
+template <IsHonkFlavor Flavor>
+void DeciderProvingKey_<Flavor>::move_structured_trace_overflow_to_overflow_block(Circuit& circuit)
+{
+    auto& blocks = circuit.blocks;
+    auto& overflow_block = circuit.blocks.overflow;
+
+    // Set has_overflow to true if a nonzero fixed size has been prescribed for the overflow block
+    blocks.has_overflow = (overflow_block.get_fixed_size() > 0);
+
+    blocks.compute_offsets(/*is_structured=*/true); // compute the offset of each fixed size block
+
+    // Check each block for capacity overflow; if necessary move gates into the overflow block
+    for (auto& block : blocks.get()) {
+        size_t block_size = block.size();
+        uint32_t fixed_block_size = block.get_fixed_size();
+        if (block_size > fixed_block_size && block != overflow_block) {
+            // We dont handle this case
+            ASSERT(!block.is_pub_inputs);
```

Comment:
why?

Reply:
update this to also include ecc_op block and added comment explaining that blocks not expected to be used by the App circuits arent allowed to overflow and will not be handled correctly